#######################
# Generate Bootstraps #
#######################

#Generate a bootstrapped alignment from the real alignment.
#The original file is also copied to the 0th bootstrap file, so that it can easily undergo the same steps as the bootstrap files (1-N).
rule generate_bootstrap_alignments:
    input:
        mafft_file="Data/{family}.mafft",
        indies_file="Data/{family}.indies"
    output:
        temp("Data/{family}.mafft_bs:{bootstrap_n}")
    shell:
        """
        set +o pipefail;
        Check_indies=`wc -l {input.indies_file} | cut -d" " -f1`
        if [ "$Check_indies" -le "2" ]; then
        	touch {output}
        else
        	if [ "{wildcards.bootstrap_n}" == "0" ]; then
        		cp Data/{wildcards.family}.mafft Data/{wildcards.family}.mafft_bs:0
        	else
        		Codes/generate_bootstrap_alignments.py {input.mafft_file} {wildcards.bootstrap_n}
        	fi
        fi
        """

#########################
# Select Two-copy Genes #
#########################

#Take only the two copy-genes.
rule extract_two_copy_genes:
    input:
        twocopies_file = "Data/{family}.mafft_bs:{bootstrap_n}",
        indies_file="Data/{family}.indies"
    output:
        temp("Data/{family}:{bootstrap_n}.tcg")
    shell:
        """
        set +o pipefail;
        Check_indies=`wc -l {input.indies_file} | cut -d" " -f1`
        if [ "$Check_indies" -le "2" ]; then
        	touch {output}
        else
        	Codes/extract_two_copy_genes.py {input.twocopies_file} {output}
        	if [ "{wildcards.bootstrap_n}" == "0" ]; then
        		rm -f Data/{wildcards.family}.mafft_bs:0
        	fi
        fi
        
        
        """

##############################
# Count Patterns in Quartets #
##############################

#Count phylogenetically informative positions (i.e. AABB, ABBA, etc.)
rule count_sequence_patterns:
    input:
        tcg_file="Data/{family}:{bootstrap_n}.tcg",
        indies_file="Data/{family}.indies"
    output:
        temp("Data/{family}:{bootstrap_n}.pts")
    shell:
    	"""
        set +o pipefail;
        Check_indies=`wc -l {input.indies_file} | cut -d" " -f1`
        if [ "$Check_indies" -le "2" ]; then
        	touch {output}
        else
        	Codes/count_sequence_patterns.py {input.tcg_file} {output}
        fi
        """

################
# Gene Network #
################

#Convert pattern counts to F (fate similarity measure) and create a network.
#Note that the networks can all be saved by removing the "temp(" and ")"; it is required in the next rule and later, to calculate the magnitude of recurrence.
#It can also be readily imported in Cytoscape.
rule make_gene_network:
    input:
        "Data/{family}:{bootstrap_n}.pts"
    output:
        temp("Data/{family}:{bootstrap_n}.fnet")
    shell:
        """
        set +o pipefail;
        Check_pts=`wc -l {input} | cut -d" " -f1`
        	if [ "$Check_pts" == "0" ]; then
        	touch {output}
        else
        	Codes/make_fate_network.py {input} {output}
        	if [ ! -f {output} ]; then
        		touch {output}
        	fi
        fi
    	"""

#################
# Cluster Fates #
#################

#Cluster the network with Markov Clustering algorithm via an R-script (requires the packages: " ").
#Current parameters: expansion = 5, inflation = 10, edge_threshold = 0.0 (no threshold).
#Genes that do no have any connections fall out of the clustering so they need to be added as separate clusters by a few lines of bash code (see beneath "shell").
rule cluster_fates:
    input:
        fate_network="Data/{family}:{bootstrap_n}.fnet",
        two_copy_genes="Data/{family}:{bootstrap_n}.tcg"
    output:
        temp("Data/{family}:{bootstrap_n}.fcl")
    shell:
        """
        set +o pipefail;
        CHECK=`wc -l {input.fate_network} | cut -d" " -f1`
        if [ "$CHECK" -lt "1" ]; then
			touch temp_{wildcards.bootstrap_n}.fcl
		else        
        	Codes/cluster_fates.R {input.fate_network} markov temp_{wildcards.bootstrap_n}.fcl 5 10 0.0
        fi
        
        cp temp_{wildcards.bootstrap_n}.fcl {output}
        
        Check_tcg2=`wc -l {input} | cut -d" " -f1`
        if [ "$Check_tcg2" == "0" ]; then
        	touch {output}
        else
        	ALL_TCG_GENES=`cat {input.two_copy_genes} | grep ">" | cut -c2-`
        	for GENE in $ALL_TCG_GENES;
        	do
            	CHECK_PRESENCE=`grep ${{GENE:0:10}} temp_{wildcards.bootstrap_n}.fcl | wc -l`
            	if [ "$CHECK_PRESENCE" -lt "1" ]; then
                	echo -en $GENE"\n" >> {output}
            	fi
        	done
        	rm -f temp_{wildcards.bootstrap_n}.fcl
        fi
        rm -f temp_{wildcards.bootstrap_n}.fcl
        """

####################
# Merge Bootstraps #
####################

#Cluster bootstraps based on jaccard scores (REF), i.e. clustering of clusters.
#Here the number of bootstraps can be set! (in the brackets of "range()").
rule cluster_bootstraps:
    input:
        original="Data/{family}:0.fcl",
        bootstraps=expand("Data/{{family}}:{bootstrap_n}.fcl", bootstrap_n=[str(x+1) for x in range(100)])
    output:
        "Data/{family}.fates"
    shell:
        """
        set +o pipefail;
        Check_original=`wc -l {input.original} | cut -d" " -f1`
        if [ "$Check_original" == "0" ]; then
        	touch {output}
        else
        	Codes/cluster_bootstraps.py {output} {input.original} {input.bootstraps}
        fi
        """

###############
# Calculate P #
###############

#Calculate the prevalence of recurrent sequence evolution (i.e P, see paper).
rule measure_recurrence_prevalence:
    input:
        indies="Data/{family}.indies",
        fates="Data/{family}.fates"
    output:
        "Data/{family}.score"
    shell:
    	"""
        set +o pipefail;
        Check_indies2=`wc -l {input.indies} | cut -d" " -f1`
        if [ "$Check_indies2" -le "2" ]; then
        	touch {output}
        	echo 'P = N/A' > {output}
		else
        	Codes/measure_recurrence_prevalence.py {input.indies} {input.fates} {output}
        fi
        """

#################
# Calculate Z_F #
#################

#First find the recurrent pairs in the largest fate of our family (i.e. the one that defines the family's P-score.
rule find_recurrent_quartets:
    input:
        indies="Data/{family}.indies",
        fates="Data/{family}.fates",
        two_copy_genes="Data/{family}:0.tcg",
        score="Data/{family}.score"
    output:
        non_rec_pairs1=temp("Data/{family}_nonrecpairs1.txt"),
        non_rec_pairs2=temp("Data/{family}_nonrecpairs2.txt"),
        recurrent_pairs=temp("Data/{family}_recurrentpairs.txt")
    shell:
        """
        set +o pipefail;
        Check_indies3=`wc -l {input.indies} | cut -d" " -f1`
        if [ "$Check_indies3" -le "2" ]; then
        	touch {output.non_rec_pairs1}
        	touch {output.non_rec_pairs2}
        	touch {output.recurrent_pairs}
		else
			Codes/find_recurrent_quartets.sh {input.indies} {input.fates} {input.two_copy_genes} {input.score} {output.non_rec_pairs1} {output.non_rec_pairs2} {output.recurrent_pairs}
        	if [ ! -f {output.recurrent_pairs} ]; then
        		touch {output.recurrent_pairs}
        	fi
        fi
    	"""
#Now calculate the magnitude of recurrent sequence evolution (i.e Z_F, see paper).
rule measure_recurrence_magnitude:
    input:
        score="Data/{family}.score",
        fates="Data/{family}.fates",
        patterns="Data/{family}:0.pts",
        non_rec_pairs1="Data/{family}_nonrecpairs1.txt",
        non_rec_pairs2="Data/{family}_nonrecpairs2.txt",
        recurrent_pairs="Data/{family}_recurrentpairs.txt"
    output:
        "Data/{family}.zscore"
    shell:
    	"""
        set +o pipefail;
        if grep -Fxq "P = N/A" {input.score}; then
        	touch {output}
        	echo 'Z = N/A' > {output}
        else
        	Codes/measure_recurrence_magnitude.sh {wildcards.family} {input.score} {input.fates} {input.patterns} {input.non_rec_pairs1} {input.non_rec_pairs2} {input.recurrent_pairs} {output}
        	if [ ! -f {output} ]; then
        		touch {output}
        		echo 'Z = N/A' > {output}
        	fi
        fi
		"""
###############################
# Identify recurrent patterns #
###############################

#Here you will generate a reordered alignment file and relating annotation files for visualization in JalView.
rule identify_recurrent_patterns:
	input:
		scores="Data/{family}.score",
		alignment="Data/{family}:0.tcg",
		indies="Data/{family}.indies",
		fates="Data/{family}.fates"
	output:
		position_colors="Output/{family}_poscol.jalview",
		groups="Output/{family}_groups.jalview",
		reordered_alignment="Output/{family}.aln",
		seq_logo="Output/{family}_seqlogo.png"
	shell:
	 	"""
        set +o pipefail;
        if grep -Fxq "P = N/A" {input.scores}; then
			touch {output.position_colors}
			touch {output.groups}
			touch {output.reordered_alignment}
			touch {output.seq_logo}
		else
		    if grep -Fxq "P = 1.0" {input.scores}; then
				touch {output.position_colors}
				touch {output.groups}
				touch {output.reordered_alignment}
				touch {output.seq_logo}
			else
        		Codes/identify_recurrent_patterns.py {input.alignment} {input.indies} {input.fates} {output.position_colors} {output.groups} {output.reordered_alignment} {output.seq_logo}
        	fi
        fi
		"""
######################
# Collect all output #
######################

rule bottom_rule:
    input:
        one="Data/{family}.score",
        two="Data/{family}.zscore",
        three="Output/{family}_poscol.jalview"
    output:
        "Output/{family}.out"
    shell:
        """
        set +o pipefail;
        cat {input.one} {input.two} > {output}
        rm -f {input.one} {input.two}
        head -1 {input.three} >> check.out
        rm -f check.out
        """
